pkg/lib/cockpit-components-dialog.jsx

```diff
@@ -209,14 +129,92 @@
  *  - id optional, id that is assigned to the top level dialog node, but not the backdrop
  *  - variant: See PF4 Modal component's 'variant' property
  *  - titleIconVariant: See PF4 Modal component's 'titleIconVariant' property
+ *  - footerProps:
+ *      - cancel_clicked optional
+ *          Callback called when the dialog is canceled
  */
 class Dialog extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            action_in_progress: false,
+            action_in_progress_promise: null,
+            action_progress_message: '',
+            action_progress_cancel: null,
+            action_canceled: false,
+        };
+        this.update_progress = this.update_progress.bind(this);
+        this.cancel_click = this.cancel_click.bind(this);
+    }
+
+    update_progress(msg, cancel) {
+        this.setState({ action_progress_message: msg, action_progress_cancel: cancel });
+    }
+
     componentDidMount() {
         // if we used a button to open this, make sure it's not focused anymore
         if (document.activeElement)
             document.activeElement.blur();
     }
 
+    action_click(handler, caption, e) {
+        this.setState({
+            action_progress_message: '',
+            action_in_progress: true,
+            action_caption_in_progress: caption,
+            action_canceled: false,
+        });
+
+        const p = handler(this.update_progress)
+                .then(() => {
+                    this.props.set_error(null);
+                    this.setState({ action_in_progress: false });
+                    if (this.props.footerProps.dialog_done)
+                        this.props.footerProps.dialog_done(true);
+                })
+                .catch(error => {
+                    if (this.state.action_canceled) {
+                        if (this.props.footerProps.dialog_done)
+                            this.props.dialog_done(false);
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19200-20230821-124816-2fdc0a03-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)